cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,229 @@
+
+    if context.logger is not None:
+        context.logger.register_initial(circuit, transformer_name)
+    circuit = func(circuit, context)
+    if context.logger is not None:
+        context.logger.register_final(circuit, transformer_name)
+    return circuit
+
+
+def transformer_method(func: TRANSFORMER_TYPE[CIRCUIT_TYPE]) -> TRANSFORMER_TYPE[CIRCUIT_TYPE]:
+    """Decorator to verify API and append logging functionality to callable transformer methods.
+
+    For example:
+    >>> @transformer_method
+    >>> def convert_to_cz(circuit: cirq.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
+    >>>    ...
+
+    Args:
+        func: Transformer implemented as a method from (CIRCUIT_TYPE, Context) --> CIRCUIT_TYPE.
+
+    Returns:
+        Method corresponding to original transformers + additional logging boilerplate.
+    """
+
+    @functools.wraps(func)
+    def transformer_with_logging_func(
+        circuit: CIRCUIT_TYPE, context: TransformerContext
+    ) -> CIRCUIT_TYPE:
+        return _transformer_with_logging(func, func.__name__, circuit, context)
+
+    return cast(TRANSFORMER_TYPE[CIRCUIT_TYPE], transformer_with_logging_func)
+
+
+def transformer_class(cls: TRANSFORMER_CLS_TYPE) -> TRANSFORMER_CLS_TYPE:
+    """Decorator to verify API and append logging functionality to callable transformer class.
+
+    The decorated class must implement the `__call__` method satisfying the `TRANSFORMER_TYPE` API.
+
+    For example:
+    >>> @transformer_class
+    >>> class ConvertToSqrtISwaps:
+    >>>    def __init__(self):
+    >>>        ...
+    >>>    def __call__(circuit: cirq.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
+    >>>        ...
+
+    Args:
+        cls: Transformer implemented as a class overriding the `__call__` method, which satisfies
+            the `TRANSFORMER_TYPE` API.
+
+    Returns:
+        `cls` with `cls.__call__` updated to include additional logging boilerplate.
+    """
+
+    old_func = cls.__call__
+
+    def transformer_with_logging_cls(
+        self: TRANSFORMER_TYPE, circuit: CIRCUIT_TYPE, context: TransformerContext
+    ) -> CIRCUIT_TYPE:
+        def call_old_func(c: CIRCUIT_TYPE, ct: TransformerContext) -> CIRCUIT_TYPE:
+            return old_func(self, c, ct)
+
+        return _transformer_with_logging(call_old_func, cls.__name__, circuit, context)
+
+    setattr(cls, '__call__', transformer_with_logging_cls)
+    return cls
+
+
+TRANSFORMER_UNION_TYPE = TypeVar(
+    'TRANSFORMER_UNION_TYPE',
+    TRANSFORMER_TYPE[frozen_circuit.FrozenCircuit],
+    TRANSFORMER_TYPE[circuit.Circuit],
+    Type[TRANSFORMER_TYPE[frozen_circuit.FrozenCircuit]],
+    Type[TRANSFORMER_TYPE[circuit.Circuit]],
+)
+
+
+@overload
+def transformer(cls_or_func: TRANSFORMER_TYPE_T) -> TRANSFORMER_TYPE_T:
+    pass
+
+
+@overload
+def transformer(cls_or_func: TRANSFORMER_CLS_TYPE) -> TRANSFORMER_CLS_TYPE:
+    pass
+
+
+def transformer(
```

Comment:
I'm not sure having a combined API for methods is all that helpful versus having a function decorator and a base class:
```python
@cirq.transformer
def my_func(circuit, context):
    ...

class MyCls(cirq.Transformer):
    def __call__(self, circuit, context):
        ...
```

Reply:
My intuition was that having a combined API can be easier to signify that we are doing the same thing in both the cases, i.e adding logging functionality to the transformer. Having two separate ways for classes and methods can potentially be confusing. 

But I don't have strong opinions here and would be happy to change if you think this is a common pattern. This would certainly make dealing with the mypy types a lot simpler.

Reply:
I guess I'm ok with the combined `cirq.transformer` call.